### PR TITLE
Add CharDevice class in qdevices

### PR DESCRIPTION
**Pull request by xutian**
CharDevice class will generate chardev device in qemu from given params dict automatically. But this version still not support hotplug and unplug, this part will finish later.

Thanks,
Xu

### CODE REVIEW

File: virttest/qemu_devices/qdevices.py [hunk 1]
```diff
@@ -960,3 +964,124 @@ def verify_unplug(self, dev_type, monitor):
         if dev_id_name not in out:
             return True
         return False
+
+
+class CharDevice(QCustomDevice):
```
Comment (spcui) [inline virttest/qemu_devices/qdevices.py]: Hi @xutian not found docs for class and function.

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -455,44 +455,63 @@ def add_serial(devices, name, filename):
...
-            cmd += _params
-            return cmd
+                port_type = 'virtserialport'
+            virt_port = QDevice(port_type)
+            virt_port.set_param("bus", bus)
+            if params.get('virtio_port_name_prefix'):
+                prefix = params["virtio_port_name_prefix"]
+                name = "%s%d" % (prefix, index)
+            virt_port.set_param("name", name)
+            virt_port.set_param("chardev", char_dev.get_aid())
```
Comment (spcui) [inline virttest/qemu_vm.py]: char_dev.get_aid() shoud be chardev.get_aid()?

**Comment by xutian** [comment]
see test results in PR
https://github.com/autotest/tp-qemu/pull/713

File: virttest/qemu_devices/qdevices.py [hunk 3]
```diff
@@ -960,3 +960,177 @@ def verify_unplug(self, dev_type, monitor):
...
+                 params=None,
+                 aobject=None,
+                 parent_bus=None,
+                 child_bus=None):
+        backend = params.get("backend", "socket")
+        self.verify_supported_backend(backend)
+        options = self.get_supported_options(backend)
+        params = params.copy_from_keys(options)
+        params = self.format_params(params)
+        params['backend'] = backend
```
Comment (CongLi) [inline virttest/qemu_devices/qdevices.py]: params['backend'] has been set in line 981.

Comment (xutian) [inline virttest/qemu_devices/qdevices.py]: no, only support backend can be store in params, see line#982, and line#981 maybe got a default values 'socket'

File: virttest/qemu_devices/qdevices.py [hunk 4]
```diff
@@ -960,3 +960,177 @@ def verify_unplug(self, dev_type, monitor):
...
         return False
+
+
+class CharDevice(QCustomDevice):
+    """
+    Qemu Char Device object, hotplug and hotunplug only support via QMP
+    monitor.
+    """
+    backends = [
+        "null", "socket", "socket", "udp", "msmouse", "vc", "ringbuf", "file",
```
Comment (CongLi) [inline virttest/qemu_devices/qdevices.py]: 2 "socket" here.

File: virttest/qemu_devices/qdevices.py (hunk 3 above)
Comment (CongLi) [inline virttest/qemu_devices/qdevices.py]: But if backend is not supported, will raise error directly.

Comment (xutian) [inline virttest/qemu_devices/qdevices.py]: right,  that's why I call verify_supported_backend()

File: virttest/qemu_devices/qdevices.py (hunk 4 above)
Comment (xutian) [inline virttest/qemu_devices/qdevices.py]: see it, thanks your comments, will update it.

**Comment by xutian** [comment]
@CongLi  updated,  would you like to review this PR please, thanks.

**Comment by CongLi** [comment]
I'm not familiar some chardev specified, but the code is good to me and test results are good.

Acked-by: Cong Li <coli@redhat.com>

**Comment by xutian** [comment]
thanks  @CongLi  and @ngu-niny 's test results.

**Comment by xutian** [comment]
thanks  @CongLi  and @ngu-niny 's test results.

**Pull request merged**